rasterio/_io.pyx

```diff
@@ -1016,43 +995,62 @@ cdef class DatasetWriterBase(DatasetReaderBase):
 
             try:
                 self._hds = exc_wrap_pointer(
-                    GDALCreate(drv, fname, self.width, self.height,
-                               self._count, gdal_dtype, options))
-            except Exception as err:
+                    GDALCreate(drv, fname, width, height,
+                               count, gdal_dtype, options))
+            finally:
                 if options != NULL:
                     CSLDestroy(options)
-                raise
 
-            if self._init_nodata is not None:
+            if nodata is not None:
 
-                if not in_dtype_range(self._init_nodata, self._init_dtype):
+                if not in_dtype_range(nodata, dtype):
                     raise ValueError(
                         "Given nodata value, %s, is beyond the valid "
                         "range of its data type, %s." % (
-                            self._init_nodata, self._init_dtype))
+                            nodata, dtype))
 
                 # Broadcast the nodata value to all bands.
-                for i in range(self._count):
+                for i in range(count):
                     band = self.band(i + 1)
-                    success = GDALSetRasterNoDataValue(band,
-                                                       self._init_nodata)
-
-            if self._transform:
-                self.write_transform(self._transform)
-            if self._crs:
-                self.set_crs(self._crs)
-            if self._init_gcps:
-                self.set_gcps(self._init_gcps, self.crs)
+                    success = GDALSetRasterNoDataValue(band, nodata)
 
-        elif self.mode == 'r+':
+        elif mode == 'r+':
             try:
                 self._hds = exc_wrap_pointer(GDALOpenShared(fname, <GDALAccess>1))
             except CPLE_OpenFailedError as err:
                 raise RasterioIOError(str(err))
 
         else:
             # Raise an exception if we have any other mode.
-            raise ValueError("Invalid mode: '%s'", self.mode)
+            raise ValueError("Invalid mode: '%s'", mode)
+
+        self.name = path
+        self.mode = mode
+        self.driver = driver
+        self.width = width
+        self.height = height
+        self._count = count
+        self._init_nodata = nodata
+        self._count = count
+        self._crs = crs
+        if transform is not None:
+            self._transform = transform.to_gdal()
+        self._gcps = None
+        self._init_gcps = gcps
+        self._closed = True
+        self._dtypes = []
+        self._nodatavals = []
+        self._units = ()
+        self._descriptions = ()
+        self._options = kwargs.copy()
+
+        if self.mode == 'w':
+            if self._transform:
+                self.write_transform(self._transform)
+            if self._crs:
+                self.set_crs(self._crs)
+            if self._init_gcps:
+                self.set_gcps(self._init_gcps, self.crs)
```

Comment:
Note: setting of dataset attributes has been moved to post-initialization of the GDAL dataset handle.